### DISCUSSION TITLE
Watchdog - stop iwdg counter in deepsleep (STOP mode)

### DISCUSSION THREAD

**Discussion by JeroenIoT** [stm32duino/Arduino_Core_STM32#1933] [Libraries] [answered]
Hi,

Have been playing with the IWatchdog and the basics are working fine.
What I want to do is the freeze the IWDG counter while in lowpower deepsleep.

I have tried the code below to freeze the counter, however it continues to reset while in deepsleep 

```
  __HAL_RCC_DBGMCU_CLK_ENABLE();
  __HAL_DBGMCU_FREEZE_IWDG();
  IWatchdog.begin(IWDG_TIMEOUT_MAX); // approx 28 seconds in my hw setup
  lp.deepSleep();                     // lowPower lib
``` 

I verified the register (DBG_IWDG_STOP at 0x4001 5808 and bit 12 is set to 1. So imho the IWDG should be frozen when in STOP mode.
What am I missing?

Chrs

**Comment by fpistm** [comment, marked as answer]
Hi,
I don't think you can stop it.
The `__HAL_DBGMCU_FREEZE_IWDG()` is used when the device enters Debug Mode (core halted) not in a low power mode.

![image](https://user-images.githubusercontent.com/20641798/215124159-da2041ca-5b9a-4466-bdf4-cf94874ff5b8.png)

As stated by the ref man:
`Once running, the IWDG cannot be stopped.`

The only way I can see is to wakeup the mcu before the iwdg expiration to reset it.

**Reply by JeroenIoT** [reply]
Thanks

Maybe you are right. I know it cannot be stopped but hoping that it was possible to freeze the IWDG counter.
Rewriting the LowPower deepsleep to wakeup e.g. every 20 seconds to reload the IWDG is not preferred :-)

I will deepdive into this during the weekend. Post my feedback later on

**Reply by fpistm** [reply]
You can freeze it the cpu is halted not in a low power mode.

**Reply by JeroenIoT** [reply]
My understanding is that it can be halted when in stop mode. 

I am a bit confused about when I actually place the mcu in stop mode and what mode I am in when placing it in deep sleep.
For me I would like to "do nothing" and wake up by an RTC alarm or by an external interrupt

**Reply by fpistm** [reply]
IWDG will not stop or freeze during low power mode. You may try to reset the mcu then after restart enter in low power without start the IWDG.

**Comment by JeroenIoT** [comment]
@fpistm - I checked the ST ref docs over the weekend and freezing the IWDG downcounter indeed seems not possible for the STM32L082. I quickly tested:

1> Go to lowpower deepsleep for 20 seconds
2> Reload IWDG counter 
3> Goto step 1 untill you slept enough

Testing the above results in an increase of approx. 1.5uA compared to not waking up to reload the IWDG
(just sharing to whomever might have same situation )

**Comment by LMESTM** [comment]
Note: The IWDG is also useful in this case to make sure you won't get stuck in sleep mode forever.
I don't know if you can change the IWDG period though

**Reply by fpistm** [reply]
> Note: The IWDG is also useful in this case to make sure you won't get stuck in sleep mode forever. I don't know if you can change the IWDG period though

Right, as stated in the README.md here:
https://github.com/stm32duino/Arduino_Core_STM32/tree/main/libraries/IWatchdog

Valid timeout values depends of the LSI clock. Typically, it is 32kH value are between
125µs and 32,768ms (~32.8 seconds). The precision depends of the timeout values:

 | timeout value range | timeout value precision |
 | ------------------- |:-----------------------:|
 | 125µs - 512ms       | 125µs
 | 513ms - 1024ms      | 250µs
 | 1025ms - 2048ms     | 500µs
 | 2049ms - 4096ms     | 1ms
 | 4097ms - 8192ms     | 2ms
 | 8193ms - 16384ms    | 4ms
 | 16385ms - 32768ms   | 8ms

**Reply by JeroenIoT** [reply]
The timeout values above are related to WWDG  
The IWDG time is "sort of" configurable up to 28 secs (afaik).
See 24.3.1 in [RefManual] (https://www.st.com/resource/en/reference_manual/rm0376-ultralowpower-stm32l0x2-advanced-armbased-32bit-mcus-stmicroelectronics.pdf

**Reply by fpistm** [reply]
Yes in the case of L0x2 as the LSI is 37kHz with a prescaler of 256 for a 12 bits counter the max value is 28.33s.
Example above is (as stated) for a LSI of 32KHz 😉
